lib/events.js

```diff
@@ -76,6 +76,74 @@ const kMaxEventTargetListeners = Symbol('events.maxEventTargetListeners');
 const kMaxEventTargetListenersWarned =
   Symbol('events.maxEventTargetListenersWarned');
 
+let EventEmitterAsyncResource;
+// The EventEmitterAsyncResource has to be initialized lazily because event.js
+// is loaded so early in the bootstrap process, before async_hooks is available.
+//
+// This implementation was adapted straight from addaleax's
+// eventemitter-asyncresource MIT-licensed userland module.
+// https://github.com/addaleax/eventemitter-asyncresource
+function lazyEventEmitterAsyncResource() {
+  if (EventEmitterAsyncResource === undefined) {
+    const {
+      AsyncResource
+    } = require('async_hooks');
+
+    const kEventEmitter = Symbol('kEventEmitter');
+    const kAsyncResource = Symbol('kAsyncResource');
+    class EventEmitterReferencingAsyncResource extends AsyncResource {
+      constructor(ee, type, options) {
+        super(type, options);
+        this[kEventEmitter] = ee;
+      }
+
+      get eventEmitter() {
+        return this[kEventEmitter];
+      }
+    }
+
+    EventEmitterAsyncResource =
+      class EventEmitterAsyncResource extends EventEmitter {
+        constructor(options = undefined) {
+          let name;
+          if (typeof options === 'string') {
+            name = options;
+            options = undefined;
+          } else {
+            name = options?.name || new.target.name;
+          }
+          super(options);
+
+          this[kAsyncResource] =
+            new EventEmitterReferencingAsyncResource(this, name, options);
+        }
+
+        emit(event, ...args) {
+          const { asyncResource } = this;
+          ArrayPrototypeUnshift(args, super.emit, this, event);
+          return ReflectApply(asyncResource.runInAsyncScope, asyncResource, args);
+        }
+
+        emitDestroy() {
+          this.asyncResource.emitDestroy();
+        }
+
+        asyncId() {
+          return this.asyncResource.asyncId();
+        }
+
+        triggerAsyncId() {
+          return this.asyncResource.triggerAsyncId();
+        }
+
+        get asyncResource() {
```

Comment:
It's somewhat inconsistent that this a getter but `asyncId`, `triggerAsyncId` are functions.

Reply:
I believe the intent here for `asyncId` and `triggerAsyncId` were to mirror the underlying `AsyncResource` API. I would have preferred getters also. What do you think @addaleax ?

Reply:
Yeah, getters for everything are also just fine imo :)

Reply:
Turned them into getters!